File: Scala3/src/main/scala/lexer/Scanner.scala
package lexer

import java.io.{ByteArrayInputStream, InputStream, InputStreamReader}
import scala.annotation.tailrec

class Scanner(val inputStreamReader: InputStreamReader):
  private var position: Int = 0
  private var current: Option[Char] = read()

  def this(inputStream: InputStream) =
    this(InputStreamReader(inputStream))

  def this(string: String) =
    this(ByteArrayInputStream(string.getBytes()))

  @tailrec
  private def read(): Option[Char] = inputStreamReader.read() match
    case -1 => None
    case i => i.toChar match
      case c if c == ' ' || c == '\r' || c == '\n' || c == '\t' => read()
      case c => Some(c)

  def peek(): Option[Char] = current

  def consume(): Option[Char] =
    val ch = peek()
    current = read()
    ch

  def isEot(): Boolean = current.isEmpty

File: Java/app/src/main/java/com/github/shin1x1/jsonparser/lexer/Scanner.java
package com.github.shin1x1.jsonparser.lexer;

import javax.annotation.Nonnull;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.Optional;

public final class Scanner {
    @Nonnull
    private final InputStreamReader reader;
    private int current;

    public Scanner(@Nonnull InputStream input) throws IOException {
        this.reader = new InputStreamReader(input);
        this.current = read();
    }

    public Scanner(@Nonnull String json) throws IOException {
        this(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)));
    }

    private int read() throws IOException {
        return reader.read();
    }

    public Optional<Character> peek() {
        return isEot() ? Optional.empty() : Optional.of((char) current);
    }

    public Optional<Character> consume() throws IOException {
        var ch = peek();

        if (!isEot()) {
            current = read();
        }

        return ch;
    }

    public boolean isEot() {
        return current == -1;
    }
}
